jetty-io/src/main/java/org/eclipse/jetty/io/ManagedSelector.java

```diff
@@ -115,28 +116,91 @@ protected void doStop() throws Exception
     {
         if (LOG.isDebugEnabled())
             LOG.debug("Stopping {}", this);
-        CloseEndPoints close_endps = new CloseEndPoints();
-        submit(close_endps);
-        close_endps.await(getStopTimeout());
-        CloseSelector close_selector = new CloseSelector();
-        submit(close_selector);
-        close_selector.await(getStopTimeout());
-
-        super.doStop();
+
+        // Grab a copy of the current selector, so we can force close endpoints later
+        Selector selector;
+        try (Locker.Lock lock = _locker.lock())
+        {
+            selector = _selector;
+        }
+
+        boolean graceful = getStopTimeout()>0;
+        boolean closed = selector==null;
+        if (graceful && !closed)
```

Comment:
If `graceful==false` I'm not sure we want to skip the submission of `GracefulCloseAction`. We still want to run that into the selector thread, just wait until it's done, no ?

Reply:
The contract for `getStopTimeout()<=0` is that `doStop` will not wait, so how can we wait until it is done?

Reply:
@gregw so it is the wait that we need to skip, not the submission of the `GracefulStopAction`, no ?

Reply:
Why submit a task if we are not going to wait for it? we are going to go close the endpoints and selectors so the submitted task would just be in a race... with latest changes it would probably lose the race and just be a noop